owntone: New package variant: owntone-pulseaudio

davidandreoletti opened the pull request:
**Maintainer**: @ejurgensen

**Compile tested**: 
 - arch: armv8
 - model: NanoPI R4S
 - OpenWrt version: 23.05
 
**Run tested**:
 - arch: armv8
 - model: NanoPI R4S
 - OpenWrt version: OpenWRT version: OpenWrt 23.05.2 r23630-842932a63d / LuCI openwrt-23.05 branch git-23.306.39416-c86c256
 - test done: 
  - [x] [owntone-pulseaudio 28.5-2](https://github.com/davidandreoletti/packages/commit/0facc7de809146bd706aad1f7d06c7aaabb04d4f) installed on NanoPi R4S
    - I compiled + tested the changes above **to** this repo's `master`. Why ? `master` uses a new version of libsodium incompatible with the lib sodium version installed on 23.05 openwrt releases. 
  - [x] setup shairport-sync to write audio to named pipe
  - [x] setup owntone read audio named pipe and stream to pulseaudio's default bluetooth sink (thanks for the write up on the owntone doc)
    - 🔊 People can hear music sent from an iPhone to OpenWRT (shairport-sync[0] -> named pipe -> owntone -> pulseaudio -> bluetooth stack) to Bluetooth Speaker !

[0] https://github.com/openwrt/packages/pull/22923
 
**Description:**
- Goal: Create package variant with pulseaudio support
- Packages size difference: + 4076 bytes
  - owntone_28.8-2_aarch64_generic.ipk: 734443 bytes
  - owntone-pulseaudio_28.8-2_aarch64_generic.ipk: 738519 bytes
  - Size reported with ls -ls --block-size 1 /tmp/shairport-sync-openssl_4.3.2-0_aarch64_generic.ipk

**Irrelevant content now:**
-  @ejurgensen Building the package "owntone-pulseaudio" fails with this [error](https://github.com/openwrt/packages/compare/master...davidandreoletti:packages:owntone-enable-pulseaudio-with-package-variant?expand=1#diff-e44cc92757151e0b7ca35593f9d70155591de9b10432f1188fac60a18dc6eaa6R79). 
    - I am hoping you can provide some guidance as to how to resolve the issues mark "FIXME".
      - @neheb's [rpath-link ](https://github.com/openwrt/packages/blob/56f3cab124f9d69e4a66ac4a610b4c2883cf5068/sound/mpd/Makefile#L199)solved the issue.

davidandreoletti committed:
owntone: Add owntone-pulseaudio package variant with pulseaudio enabled

Packages size difference: +4076 bytes
- owntone_28.8-2_aarch64_generic.ipk: 734443 bytes
- owntone-pulseaudio_28.8-2_aarch64_generic.ipk: 738519 bytes

Size reported with 'ls -ls --block-size 1 /tmp/shairport-sync-openssl_4.3.2-0_aarch64_generic.ipk'

Signed-off-by: David Andreoletti <david@andreoletti.net>

neheb commented:
If I had to guess, rpath-link hack as used in mpd is needed.

davidandreoletti commented:
> If I had to guess, rpath-link hack as used in mpd is needed.

@neheb Good guess! Thank you.

davidandreoletti commented:
`owntone-pulseaudio` is defined with `DEPENDS+= +AUDIO_SUPPORT:pulseaudio`. 
- `pulseaudio` is a virtual package, declared in:
  - `pulseaudio-daemon`
  - `pulseaudio-daemon-avahi`

On my router:
 0. `pulseaudio-daemon-avahi` is installed
 1  `opkg install /tmp/owntone-pulseaudio_28.8-2_aarch64_generic.ipk` fails:
  
```
Installing owntone-pulseaudio (28.8-2) to root...
Installing pulseaudio-daemon (14.2-10) to root...
Collected errors:
 * check_conflicts_for: The following packages conflict with pulseaudio-daemon:
 * check_conflicts_for:         pulseaudio-daemon-avahi *
 * opkg_install_cmd: Cannot install package owntone-pulseaudio.
```

@neheb How to specify `owntone-pulseaudio` is fine with either dependency `pulseaudio-daemon` or `pulseaudio-daemon-avahi` installed, as long as one of them installed ?

davidandreoletti commented:
Moving the discussion to https://forum.openwrt.org/t/owntone-support-pulseaudio/182872?u=dandreoletti

davidandreoletti commented:
@ejurgensen 
 - Should this PR be merged in https://github.com/owntone/owntone-openwrt/tree/master instead ?
 - Let me know if further work needed on the PR itself.

ejurgensen commented:
> Should this PR be merged in https://github.com/owntone/owntone-openwrt/tree/master instead ?

I think here is the right place, the owntone repo is for testing/CI. You are welcome to make PR to that repo too, but otherwise I will pull it myself from here.

> Let me know if further work needed on the PR itself.

Looks good to me, but I should say that I'm not an OpenWrt packaging expert.

> People can hear music sent from an iPhone to OpenWRT (shairport-sync[0] -> named pipe -> owntone -> pulseaudio -> bluetooth stack) to Bluetooth Speaker !

Super cool!

BKPepe commented:
1. Commit description is missing.

Ad) the size difference. Does it make sense to create variant in the first place instead of enabling it by default? Doing that for +4076 bytes. Hmmm. I would rather see that enabled it by default. You know. Every year requirements are higher and higher. Software is using more storage, etc.

davidandreoletti commented:
> I will pull it myself from here.

Ok

davidandreoletti commented:
> Commit description is missing.

Got it. Will update later

davidandreoletti commented:
>Does it make sense to create variant in the first place instead of enabling it by default? Doing that for +4076 bytes. Hmmm. I would rather see that enabled it by default.

I agree: Enabled by default. With ``` DEPENDS+= +AUDIO_SUPPORT:pulseaudio```, targets not supporting audio support will gracefully compile without pulseaudio support, ie current `owntone` package.

davidandreoletti committed:
owntone: Enable pulseaudio support on config with audio support enabled

Packages size difference: +4076 bytes
- owntone_28.8-2_aarch64_generic.ipk: 734443 bytes
- owntone-pulseaudio_28.8-2_aarch64_generic.ipk: 738519 bytes

PS: Size reported with 'ls -ls --block-size 1 /tmp/shairport-sync-openssl_4.3.2-0_aarch64_generic.ipk'
Signed-off-by: David Andreoletti <david@andreoletti.net>

davidandreoletti commented:
> > Does it make sense to create variant in the first place instead of enabling it by default? Doing that for +4076 bytes. Hmmm. I would rather see that enabled it by default.
> 
> I agree: Enabled by default. With ` DEPENDS+= +AUDIO_SUPPORT:pulseaudio`, targets not supporting audio support will gracefully compile without pulseaudio support, ie current `owntone` package.
> 
> The PR has 2 commits:
> 
> * (a) [b46a8fe](https://github.com/openwrt/packages/commit/b46a8fe8a43356c5abead46869d81911d2e496b0): new `owntone-pulseaudio` package variant
> * (b) [194d4c8](https://github.com/openwrt/packages/commit/194d4c86b9e66aa58d5dc3265b902983496a52a7): original `owntone` package with conditional pulseaudio support added
> 
> @BKPepe / @ejurgensen Once a decision is taken, I will remove the irrelevant commit.
> 
> I prefer (b).

@BKPepe / @ejurgensen (ping)

davidandreoletti commented:
This is how I understood things.

Each OpenWRT device has different audio capabilities. 

AUDIO_SUPPORT is there to enable building against pulseaudio for this owntone package only iff the kernel / .config file support/enable it.

For my particular case, I did not need it and think it is relevant for other device targets.

Without this conditional package dependency inclusion statement, owntone would fail to build for some other device targets.

davidandreoletti commented:
> @BKPepe Would mind amending gaps in my understanding, if any please ?

(Ping @BKPepe)

BKPepe commented:
I will look at it. It looks like we dont have enough reviewers and I am trying to do it as much as I can in my free time, which I dont have much these days.

davidandreoletti commented:
> I will look at it. It looks like we dont have enough reviewers and I am trying to do it as much as I can in my free time, which I dont have much these days.

@BKPepe 

I would like to merge this (once ready) before the next OpenWRT stable release. 

So far the next release  ([timeframe seems not defined in any new meetings](https://openwrt.org/meetings/start)) and [no release target set either](https://github.com/orgs/openwrt/projects).

Does that give you more or less time to review + answer[ this question](https://github.com/openwrt/packages/pull/22942#issuecomment-1907189412) ?

GeorgeSapkin commented:
@davidandreoletti still relevant? There are merge conflicts.

davidandreoletti committed:
Merge branch 'master' into owntone-enable-pulseaudio-with-package-variant

Signed-off-by: David Andreoletti <david@andreoletti.net>

davidandreoletti commented:
> @davidandreoletti still relevant? There are merge conflicts.

Not relevant for me at the moment yet PR updated if anyone else needs it.
